### PR TITLE
Add ParentClosePolicy

### PR DESCRIPTION
**Pull request by tim37021**
I found [`ParentClosePolicy`](https://docs.temporal.io/docs/content/what-is-a-parent-close-policy/) is not supported. Here's the PR.

## Story
I am implementing proxy workflow pattern mention [HERE](https://community.temporal.io/t/sync-wait-a-workflow/3033)
I want to launch a child workflow and the main workflow will finish earlier then it's child. Due to default parent close policy. The child workflow will be killed as well. 

### without `parent_close_policy`
image

### with `parent_close_policy: :abandon`
image

## Usage
```ruby
class ProxyWorkflow < Temporal::Workflow
  def execute(workflow_id)
    # copied from SDK example
    future = LongRunningActivity.execute(10, 1)
    workflow.execute_workflow(
      MainWorkflow,
      proxy_workflow_id: workflow_id,
      proxy_run_id: workflow.metadata.run_id,
      options: {
        parent_close_policy: :abandon,
      },
    )
    ret = nil
    workflow.on_signal do |signal, input|
      future.cancel
      ret = input
    end
    future.wait
    ret
  end
end
```

```ruby
class MainWorkflow < Temporal::Workflow
  def execute(proxy_workflow_id:, proxy_run_id:)
    workflow.sleep(5)
    SignalActivity.execute!(proxy_workflow_id: proxy_workflow_id, proxy_run_id: proxy_run_id, input: "succsss yeah")
    10.times do
      HelloActivity.execute!("hehehe")
      workflow.sleep(1)
    end
  end
end
```

By the way, how can I know my workflow id? Currently I explicitly names it.
```ruby
Temporal.start_workflow(TemporalWorkflow::ProxyWorkflow,  "hehehe", options: {workflow_id: "hehehe"})
```

### CODE REVIEW

File: lib/temporal/parent_close_policy.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    class InvalidParentClosePolicy < ClientError; end
+
+    attr_reader :policy
+
+    def initialize(policy)
+      @policy = policy
+    end
+
+    def validate!
+      unless %i[terminate abandon request_cancel].include?(@policy)
```
Comment (antstorm) [inline lib/temporal/parent_close_policy.rb:15]: Can you please define these as constants so they can be used without hardcoding symbols?

File: lib/temporal/execution_options.rb [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
-    attr_reader :name, :namespace, :task_queue, :retry_policy, :timeouts, :headers
+    attr_reader :name, :namespace, :task_queue, :retry_policy, :parent_close_policy, :timeouts, :headers
 
     def initialize(object, options, defaults = nil)
       # Options are treated as overrides and take precedence
       @name = options[:name] || object.to_s
       @namespace = options[:namespace]
       @task_queue = options[:task_queue] || options[:task_list]
       @retry_policy = options[:retry_policy] || {}
+      @parent_close_policy = options[:parent_close_policy] || {}
```
Comment (antstorm) [inline lib/temporal/execution_options.rb:15]: ExecutionOptions are shared between workflows and activities, so this wouldn't be the best place to add workflow-specific options. Can you please move these inline with the child workflow invocation?

File: spec/unit/lib/temporal/connection/serializer/parent_close_policy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      end.not_to raise_error
+
+      expect do
+        described_class.new(abandon_policy).to_proto
+      end.not_to raise_error
+
+      expect do
+        described_class.new(request_cancel_policy).to_proto
+      end.not_to raise_error
+    end
```
Comment (antstorm) [inline spec/unit/lib/temporal/connection/serializer/parent_close_policy_spec.rb:22]: Can you please split these into 3 examples? It's easier to figure out which ones have failed that way

**Comment by tim37021** [comment]
> Good call on getting the workflow_id when starting a workflow without explicitly providing it, we don't have a way to do that currently. 
I think it's possible to not to have break change? instead of changing the interface.

I remember we have a `run_id` inside `workflow.metadata`
but it doesn't contains `workflow_id` currently. is it good to add that inside metadata?

**Comment by antstorm** [comment]
@tim37021 yes, I was thinking that the most logical solution would be to return both workflow_id and run_id for a `start_workflow` call:

```ruby
workflow_id, run_id = Temporal.start_workflow(MyWorkflow)
```

However exposing this via metadata still works, but I'm not sure how you're planning to get the metadata.

As for adding workflow_id to metadata — yes, that's something I've already implemented for cadence — https://github.com/coinbase/cadence-ruby/pull/60, I'll replay that change over this repo today since others have asked about it as well

**Comment by antstorm** [comment]
There we go — https://github.com/coinbase/temporal-ruby/pull/110

**Pull request closed without merging**